File: testers/sqlite/java/src/java/testdb/BasicCrudTest.java
package testdb;

import static org.junit.Assert.*;

import java.util.List;
import org.junit.Test;
import testdb.customers.*;
import testdb.products.*;

/** Minimal end-to-end SQLite test: inserts, selects, updates, deletes. */
public class BasicCrudTest {
  private final CustomersRepoImpl customers = new CustomersRepoImpl();
  private final ProductsRepoImpl products = new ProductsRepoImpl();

  @Test
  public void selectAllCustomersFromSeedData() {
    List<CustomersRow> rows = SqliteTestHelper.applyRead(customers::selectAll);
    assertEquals(2, rows.size());
  }

  @Test
  public void findCustomerById() {
    var row = SqliteTestHelper.applyRead(conn -> customers.selectById(new CustomersId(1L), conn));
    assertTrue(row.isPresent());
    assertEquals("John Doe", row.get().name());
  }

  @Test
  public void deleteCustomerById() {
    // Insert a fresh customer (no FK references) and delete it within the same transaction.
    boolean deleted =
        SqliteTestHelper.apply(
            conn -> {
              var inserted =
                  customers.insert(
                      new CustomersRowUnsaved(
                          new CustomersId(999L),
                          "Disposable",
                          java.util.Optional.empty(),
                          new testdb.customtypes.Defaulted.UseDefault<>()),
                      conn);
              return customers.deleteById(inserted.customerId(), conn);
            });
    assertTrue(deleted);
  }

  @Test
  public void productsHaveExpectedPrices() {
    var ps = SqliteTestHelper.applyRead(products::selectAll);
    assertEquals(2, ps.size());
    var skus = ps.stream().map(ProductsRow::sku).sorted().toList();
    assertEquals(List.of("PROD-001", "PROD-002"), skus);
  }
}

File: testers/sqlite/kotlin/src/kotlin/testdb/BasicCrudTest.kt
package testdb

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertTrue
import org.junit.Test
import testdb.customers.CustomersId
import testdb.customers.CustomersRepoImpl
import testdb.customers.CustomersRowUnsaved
import testdb.customtypes.Defaulted
import testdb.products.ProductsRepoImpl

class BasicCrudTest {
    private val customers = CustomersRepoImpl()
    private val products = ProductsRepoImpl()

    @Test
    fun selectAllCustomersFromSeedData() {
        val rows = SqliteTestHelper.applyRead { customers.selectAll(it) }
        assertEquals(2, rows.size)
    }

    @Test
    fun findCustomerById() {
        val row = SqliteTestHelper.applyRead { customers.selectById(CustomersId(1L), it) }
        assertNotNull(row)
        assertEquals("John Doe", row!!.name)
    }

    @Test
    fun deleteCustomerById() {
        val deleted = SqliteTestHelper.apply { conn ->
            val inserted = customers.insert(
                CustomersRowUnsaved(CustomersId(999L), "Disposable", null, Defaulted.UseDefault()), conn
            )
            customers.deleteById(inserted.customerId, conn)
        }
        assertTrue(deleted)
    }

    @Test
    fun productsHaveExpectedPrices() {
        val ps = SqliteTestHelper.applyRead { products.selectAll(it) }
        assertEquals(2, ps.size)
        val skus = ps.map { it.sku }.sorted()
        assertEquals(listOf("PROD-001", "PROD-002"), skus)
    }
}
